numpy/core/src/npymath/npy_math_internal.h.src

```diff
@@ -900,3 +902,390 @@ npy_popcount@c@(npy_@type@ a)
+    si++;
+    m -= (int)m;
+    s[si] = '.';
+    si++;
+    for (i=0; i < (TOHEX_NBITS-1)/4; i++) {
+        m *= 16.0;
+        s[si] = char_from_hex((int)m);
+        si++;
+        m -= (int)m;
+    }
+    s[si] = '\0';
+
+    if (e < 0) {
+        esign = (int)'-';
+        e = -e;
+    } else {
+        esign = (int)'+';
+    }
+
+    if (x < 0.0) {
+        sprintf(buff, "-0x%sp%c%d", s, esign, e);
+    } else {
+        sprintf(buff, "0x%sp%c%d", s, esign, e);
+    }
+}
+
+/* Heavily copied from the builtin float.fromhex
+ * This function returns True if conversion is possible,
+ * else returns false and value of the argument `result`
+ * should not be used.
+ */
+NPY_INPLACE npy_bool
+npy_@name@_fromhex(const char *string, npy_@name@ *result)
+{
+    npy_@name@ x;
+    long exp, top_exp, lsb, key_digit;
+    const char *s, *coeff_start, *s_store, *coeff_end, *exp_start, *s_end;
+    npy_int half_eps, digit, round_up, negate=0;
+    Py_ssize_t ndigits, fdigits, i;
+
+    /*
+     * For the sake of simplicity and correctness, we impose an artificial
+     * limit on ndigits, the total number of hex digits in the coefficient
+     * The limit is chosen to ensure that, writing exp for the exponent,
+     *
+     *   (1) if exp > LONG_MAX/2 then the value of the hex string is
+     *   guaranteed to overflow (provided it's nonzero)
+     *
+     *   (2) if exp < LONG_MIN/2 then the value of the hex string is
+     *   guaranteed to underflow to 0.
+     *
+     *   (3) if LONG_MIN/2 <= exp <= LONG_MAX/2 then there's no danger of
+     *   overflow in the calculation of exp and top_exp below.
+     *
+     * More specifically, ndigits is assumed to satisfy the following
+     * inequalities:
+     *
+     *   4*ndigits <= DBL_MIN_EXP - DBL_MANT_DIG - LONG_MIN/2
+     *   4*ndigits <= LONG_MAX/2 + 1 - DBL_MAX_EXP
+     *
+     * If either of these inequalities is not satisfied, a ValueError is
+     * raised.  Otherwise, write x for the value of the hex string, and
+     * assume x is nonzero.  Then
+     *
+     *   2**(exp-4*ndigits) <= |x| < 2**(exp+4*ndigits).
+     *
+     * Now if exp > LONG_MAX/2 then:
+     *
+     *   exp - 4*ndigits >= LONG_MAX/2 + 1 - (LONG_MAX/2 + 1 - DBL_MAX_EXP)
+     *                    = DBL_MAX_EXP
+     *
+     * so |x| >= 2**DBL_MAX_EXP, which is too large to be stored in C
+     * double, so overflows.  If exp < LONG_MIN/2, then
+     *
+     *   exp + 4*ndigits <= LONG_MIN/2 - 1 + (
+     *                      DBL_MIN_EXP - DBL_MANT_DIG - LONG_MIN/2)
+     *                    = DBL_MIN_EXP - DBL_MANT_DIG - 1
+     *
+     * and so |x| < 2**(DBL_MIN_EXP-DBL_MANT_DIG-1), hence underflows to 0
+     * when converted to a C double.
+     *
+     * It's easy to show that if LONG_MIN/2 <= exp <= LONG_MAX/2 then both
+     * exp+4*ndigits and exp-4*ndigits are within the range of a long.
+     */
+
+    s = string;
+    if (s == NULL)
+        return NPY_FALSE;
+    s_end = s + strlen(s);
```

Comment:
This logic is incorrect as we can inject a `\0` as shown [here](https://github.com/numpy/numpy/pull/20083/files#diff-c5ebcf9e35c56cb23da228c4bea6471c8587296ea82c5a0ee3da5fe3f6dbc741R287).

I'll try to compute length in a different way or worst case, pass it an arg to this function or maybe pass the start and end of the string in an iterator style.

Reply:
Handled in 2ba28ca